File: apps/relay-go/pkg/engine/pot.go
// Package engine is the pure deterministic In-Between state machine.
// Every exported function is total over (state, action, ruleSet, height)
// and returns either the successor state or a typed ProtocolError.
// There is no clock, no I/O, no shared state.
package engine

import (
	"fmt"
	"math"

	"github.com/prof-faustus/cardtable/relay-go/pkg/types"
)

// AddToPot adds delta satoshis to the pot. Overflow into the unsafe
// integer range raises an error.
func AddToPot(pot, delta types.Satoshis) (types.Satoshis, error) {
	next := uint64(pot) + uint64(delta)
	if next > math.MaxInt64 {
		return 0, fmt.Errorf("AddToPot: pot overflow (%d + %d)", pot, delta)
	}
	return types.Satoshis(next), nil
}

// SubtractFromPot subtracts delta from the pot. Underflow raises.
func SubtractFromPot(pot, delta types.Satoshis) (types.Satoshis, error) {
	if delta > pot {
		return 0, fmt.Errorf("SubtractFromPot: underflow (%d - %d)", pot, delta)
	}
	return pot - delta, nil
}

File: packages/state-engine/src/pot.ts
/**
 * Pot accounting.
 *
 * Pure functions over satoshi amounts. No I/O, no clock.
 */

import type { Satoshis } from '@cardtable/protocol-types';
import { asSatoshis } from '@cardtable/protocol-types';

/**
 * Add `delta` satoshis to the pot. Both inputs must be non-negative.
 * Throws on overflow into unsafe-integer range.
 */
export function addToPot(pot: Satoshis, delta: Satoshis): Satoshis {
  const next = pot + delta;
  if (!Number.isSafeInteger(next)) {
    throw new Error(`addToPot: pot overflow (${pot} + ${delta})`);
  }
  return asSatoshis(next);
}

/**
 * Subtract `delta` satoshis from the pot. Throws if delta exceeds pot.
 */
export function subtractFromPot(pot: Satoshis, delta: Satoshis): Satoshis {
  if (delta > pot) {
    throw new Error(`subtractFromPot: underflow (${pot} - ${delta})`);
  }
  return asSatoshis(pot - delta);
}
